### PR TITLE
Improvement/zenko 4617 integrate kfkacleaner

### PR DESCRIPTION
**Pull request by KillianG**
This pull request primarily introduces changes to integrate a Kafka cleaner and to enhance the testing of Kafka's consumed messages. The Kafka cleaner is a new component added to the Zenko solution, which is reflected in the configuration files. The testing enhancements include a new test case and additional testing libraries.

Integration of Kafka Cleaner:

* [`.github/scripts/end2end/configs/zenko.yaml`](diffhunk://#diff-6b63bfb71ed8974ce525e53e0a41053e7b3f52153bdeab39ebc1c82ee488a5e4R52-R53): Added a Kafka cleaner with a specified interval.
* [`.github/scripts/end2end/configs/zenkoversion.yaml`](diffhunk://#diff-6e6088aa473aa9d62926825912523b2e33d766727118ca0d910bfc8de0f98152R102-R104): Added image and tag details for the Kafka cleaner.
* [`solution/deps.yaml`](diffhunk://#diff-1f2b6609225d72aca60548ff66cb69731669e72901ad8a358a180465f1538836R36-R40): Added the Kafka cleaner with sourceRegistry, image, tag, and envsubst details.
* [`solution/zenkoversion.yaml`](diffhunk://#diff-81f8312b28c4dda5038fc634db8687463d58b1c87d41b5c0f2a791f99cc6c951R21-R23): Added the Kafka cleaner to the Zenko version configuration with image and tag details. [[1]](diffhunk://#diff-81f8312b28c4dda5038fc634db8687463d58b1c87d41b5c0f2a791f99cc6c951R21-R23) [[2]](diffhunk://#diff-81f8312b28c4dda5038fc634db8687463d58b1c87d41b5c0f2a791f99cc6c951R103-R105)

Enhancements to Kafka Testing:

* [`.github/scripts/end2end/run-e2e-ctst.sh`](diffhunk://#diff-a8a4d241ff23517544a6ae0ddf6d27875543114739b9ba34b36fd0c7f270e78aR57-R61): Added an INSTANCE_ID to the world parameters.
* [`tests/ctst/common/common.ts`](diffhunk://#diff-e6ae52857aee30890b3ab95b85a62914435d01c1c53df945c0dce3cf8fe7742eL1-R18): Imported additional libraries and added a new function to get topic offsets, and a new test case to verify that consumed Kafka messages do not take up too much disk space. [[1]](diffhunk://#diff-e6ae52857aee30890b3ab95b85a62914435d01c1c53df945c0dce3cf8fe7742eL1-R18) [[2]](diffhunk://#diff-e6ae52857aee30890b3ab95b85a62914435d01c1c53df945c0dce3cf8fe7742eR80-R101)
* [`tests/ctst/features/zzz.kafkaCleaner.feature`](diffhunk://#diff-12d2ae0444e5bfa89e124c9a428308c912225b515a2f045b15dc2a28a5282eecR1-R9): Added a new feature file to test the Kafka cleaner.
* [`tests/ctst/package.json`](diffhunk://#diff-38733078a0ef869cbfaf7770fc526c881d0cae93724aba1983ae5761ceaa3760R14): Added the KafkaJS library to the package dependencies.
* [`tests/ctst/world/Zenko.ts`](diffhunk://#diff-371f6719b5883f728d9d3218b4ed2eaf1b7ada8faf798acf95b686a7c347c7ddR86): Added an InstanceID to the ZenkoWorldParameters interface.

### CODE REVIEW

File: solution/zenkoversion.yaml [hunk 1]
```diff
@@ -97,6 +100,9 @@ spec:
       monitoring:
         image: jmx-javaagent
         tag: '${JMX_JAVAAGENT_TAG}'
+      cleaner:
+        image: kafka-cleaner-dev
```
Comment (welansari) [inline solution/zenkoversion.yaml]:
```suggestion
        image: kafka-cleaner
```

File: solution/zenkoversion.yaml [hunk 2]
```diff
@@ -18,6 +18,9 @@ spec:
     kafkaConnect:
       image: kafka-connect-dashboard
       tag: '${VERSION_FULL}'
+    kafkaCleaner:
+      image: kafka-cleaner-dashboard
+      tag: '${VERSION_FULL}'
```
Comment (welansari) [inline solution/zenkoversion.yaml]: nit: should keep alphabetical order ?
same for `cleaner` in the same file

File: solution/zenkoversion.yaml (hunk 1 above)
Comment (KillianG) [reply solution/zenkoversion.yaml]: thats done, was waiting for the release of the component

File: tests/ctst/common/common.ts [hunk 3]
```diff
@@ -66,3 +78,27 @@ Given('an object {string} that {string}',
...
+Then('kafka consumed messages should not take too much place on disk',
+    async function (this: Zenko) {
+        const notToCheckTopics = ['oplog', 'dead-letter'];
+        const kafkaAdmin = new Kafka({ brokers: [this.parameters.KafkaHosts] }).admin();
+        const topics: string[] = (await kafkaAdmin.listTopics())
+            .filter(t => (t.includes(this.parameters.InstanceID) &&
+            !notToCheckTopics.some(e => t.includes(e))));
+        const previousOffsets = await getTopicsOffsets(topics, kafkaAdmin);
+
+        await Utils.sleep(35000);
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: what is this delay for / related to?

* this definitely needs a comment to explain
* ideally, we don't want a hard-coded value but to derive/sync this value with its 'root cause' : e.g. either automatically read KFKC period from the cluster, or allow the delay to be configured to match

File: solution/deps.yaml [hunk 4]
```diff
@@ -28,6 +28,11 @@ kafka:
   image: kafka
   tag: 2.13-3.1.2
   envsubst: KAFKA_TAG
+kafka-cleaner:
```
Comment (francoisferrand) [inline solution/deps.yaml]: already bump'ed, should be removed from here

File: tests/ctst/common/common.ts [hunk 5]
```diff
@@ -66,3 +78,32 @@ Given('an object {string} that {string}',
...
+        const kafkaAdmin = new Kafka({ brokers: [this.parameters.KafkaHosts] }).admin();
+        const topics: string[] = (await kafkaAdmin.listTopics())
+            .filter(t => (t.includes(this.parameters.InstanceID) &&
+            !notToCheckTopics.some(e => t.includes(e))));
+        const previousOffsets = await getTopicsOffsets(topics, kafkaAdmin);
+
+        const seconds = parseInt(this.parameters.KafkaCleanerInterval);
+
+        // Waiting for kafkacleaner to run
+        await Utils.sleep(seconds * 1000);
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: then delay should be based on KAFKA_CLEANER_INTERVAL ?

File: tests/ctst/common/common.ts [hunk 6]
```diff
@@ -66,3 +78,32 @@ Given('an object {string} that {string}',
             ));
         }
     });
+
+Then('kafka consumed messages should not take too much place on disk',
+    async function (this: Zenko) {
+        const notToCheckTopics = ['oplog', 'dead-letter'];
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: "topics not to check" → `topicsNotToCheck` (or `topicsToIgnore`)
or the other way with an adjective : `ignoredTopics`

Comment (francoisferrand) [inline tests/ctst/common/common.ts]: i don't think we should ignore `oplog` : it should be cleaned up after it has been consumed...

File: tests/ctst/common/common.ts [hunk 7]
```diff
@@ -66,3 +78,32 @@ Given('an object {string} that {string}',
...
+        for (let i = 0; i < topics.length; i++) {
+            process.stdout.write(`\nChecking topic ${topics[i]}\n`);
+            for (let j = 0; j < newOffsets[i].partitions.length; j++) {
+                // Checking that the min offset has increased due to kafkacleaner
+                // or that it didn't need to change because there was no new messages
+                assert.ok(newOffsets[i].partitions[j].low > previousOffsets[i].partitions[j].low ||
+                    newOffsets[i].partitions[j].high === newOffsets[i].partitions[j].low,
+                `Topic ${topics[i]} partition ${j} low offset has not increased, newOffsets:
+                    ${newOffsets[i].partitions[j].low}, previousOffsets: ${previousOffsets[i].partitions[j].low}`);
+            }
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: For most topic, the partition should be empty.

There may only be a very few number of topcs which remain active after the tests: bucket-topics for sure (as the conductor will keep polling), but should not mcuh else (if anything).
→ would be better to check strictly, and accept that the topic is empty only if new messages have indeed been pushed in the meantime only for these few topics

File: tests/ctst/common/common.ts [hunk 8]
```diff
@@ -66,3 +78,39 @@ Given('an object {string} that {string}',
...
+        for (let i = 0; i < topics.length; i++) {
+            process.stdout.write(`\nChecking topic ${topics[i]}\n`);
+            for (let j = 0; j < newOffsets[i].partitions.length; j++) {
+                // Checking that the min offset has increased due to kafkacleaner
+                // or that it didn't need to change because there was no new messages
+                assert.ok(newOffsets[i].partitions[j].low > previousOffsets[i].partitions[j].low ||
+                    newOffsets[i].partitions[j].high === newOffsets[i].partitions[j].low,
+                `Topic ${topics[i]} partition ${j} offset has not increased,
+                previousOffsets: ${previousOffsets[i].partitions[j].low} / ${previousOffsets[i].partitions[j].high},
+                newOffsets: ${newOffsets[i].partitions[j].low} / ${newOffsets[i].partitions[j].high}`);
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts:113]: nit: indent
```suggestion
                    `Topic ${topics[i]} partition ${j} offset has not increased,
                    previousOffsets: ${previousOffsets[i].partitions[j].low} / ${previousOffsets[i].partitions[j].high},
                    newOffsets: ${newOffsets[i].partitions[j].low} / ${newOffsets[i].partitions[j].high}`
                );
```

**Review by francoisferrand — APPROVED** [review]

**Review by welansari — APPROVED** [review]

**Pull request merged**